### PR TITLE
raftstore-v2: supply SlowTrend for v2 to detect network or disk ios' jitters

### LINKED ISSUE #14750: raftstore-v2: adaptive to SlowTrend for detecting slowness.
## Development Task
In previous work, we've already replace `SlowScore` with `SlowTrend` for detecting the disk-io jitters.
For HA, we should make it adaptive to `raftstore-v2`.

### PR DESCRIPTION
**Pull request by LykxSassinator**
### What is changed and how it works?

Issue Number: Close #14750 

What's Changed:

```commit-message
Supply `SlowTrend` in raftstore-v2 for detecting network or disk ios' jitters.
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None.
```

### CODE REVIEW

File: components/raftstore-v2/src/worker/pd/slowness.rs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        // Record IO latency on Disk.
+        self.slowness_stats.slow_cause.record(
+            tikv_util::time::duration_to_us(duration.sum()),
+            Instant::now(),
+        );
+    }
+
+    pub fn handle_slowness_stats_tick(&mut self) {
+        // Record a fairly great value when tick
+        self.slowness_stats
```
Comment (innerr) [inline components/raftstore-v2/src/worker/pd/slowness.rs:83]: This is confusing to me, we should record a greate value only when (the inspecting request is) timeout, but current code will record 100ms at every tick.

I understand that this (data in slowness) won't be sent if the store heartbeat to pd is un-delayed, but the data set in slowness still be abnormal.

File: components/raftstore-v2/src/worker/pd/store.rs [hunk 2]
```diff
@@ -14,13 +14,23 @@ use pd_client::{
...
+    time::{Instant as TiInstant, UnixSecs},
+    topn::TopN,
+};
 
 use super::Runner;
+use crate::router::StoreMsg;
 
 const HOTSPOT_REPORT_CAPACITY: usize = 1000;
 
+/// Max limitation of delayed store_heartbeat.
```
Comment (innerr) [inline components/raftstore-v2/src/worker/pd/store.rs:31]: Put a time unit or a value explaination (5min) in the comment would be nice.

File: components/raftstore-v2/src/worker/pd/slowness.rs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: PdClient + 'static,
+{
+    /// Record slowness periodically.
+    pub fn handle_update_slowness_stats(&mut self, _tick: u64, duration: RaftstoreDuration) {
+        self.slowness_stats.last_tick_finished = true;
+        // Record IO latency on Disk.
+        self.slowness_stats.slow_cause.record(
+            tikv_util::time::duration_to_us(duration.sum()),
```
Comment (innerr) [inline components/raftstore-v2/src/worker/pd/slowness.rs:76]: Not sure about this: the sum duration is complicated (comparing to the original IO-only one), is it work well in tests?

File: components/raftstore-v2/src/worker/pd/slowness.rs (hunk 1 above)
Comment (LykxSassinator) [reply components/raftstore-v2/src/worker/pd/slowness.rs:83]: Yep, i've noticed this point. Originally, I wanna use a periodical tick(100ms) to make it tolerable to minor io jitters, but it's really confusing to others.

I'm considering whether to supplement more annotations or remove it.

File: components/raftstore-v2/src/worker/pd/slowness.rs (hunk 3 above)
Comment (LykxSassinator) [reply components/raftstore-v2/src/worker/pd/slowness.rs:76]: Yep, it's a proved modification according to the significant results proved by tests on scenarios on Disk IO jitters and Network IO jitters.

Comment (LykxSassinator) [reply components/raftstore-v2/src/worker/pd/slowness.rs:76]: And as discussed offline, it's more appropriate to divide it into two individual parts (factors) for further development if possible.

File: components/raftstore-v2/src/batch/store.rs [hunk 4]
```diff
@@ -273,6 +277,27 @@ impl<EK: KvEngine, ER: RaftEngine, T: Transport + 'static> PollHandler<PeerFsm<E
...
     fn end(&mut self, _batch: &mut [Option<impl DerefMut<Target = PeerFsm<EK, ER>>>]) {
         let dur = self.timer.saturating_elapsed();
+
+        let mut latency_inspect = std::mem::take(&mut self.poll_ctx.pending_latency_inspect);
+        for inspector in &mut latency_inspect {
+            inspector.record_store_process(dur);
+        }
+        // Use the valid size of async-ios for generating `writer_id` when the local
+        // senders haven't been updated by `poller.begin().
+        let writer_id = rand::random::<usize>()
```
[the remark this answers is not in the document]
Comment (LykxSassinator) [reply components/raftstore-v2/src/batch/store.rs:287]: Not recommended. The `Config::raftstore-io-pool-size` might be modified by users, and we should ensure the new message will be randomly sent to the sender correctly.

File: components/raftstore-v2/src/worker/pd/store.rs (hunk 2 above)
Comment (LykxSassinator) [reply components/raftstore-v2/src/worker/pd/store.rs:31]: Thx.

**Review by innerr — APPROVED** [review]

File: components/raftstore-v2/src/worker/pd/slowness.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            // FAKE `store-heartbeat`. It's an assurance that the heartbeat to
+            // PD is not lost. Normally, this case rarely happens in
+            // raftstore-v2.
+            self.handle_fake_store_heartbeat();
+        }
+        // Move to next tick.
+        self.slowness_stats.last_tick_finished = false;
+    }
+
+    pub fn update_slowness_in_store_stats(&mut self, stats: &mut pdpb::StoreStats, query_num: u64) {
```
Comment (tonyxuqqi) [inline components/raftstore-v2/src/worker/pd/slowness.rs:104]: It's better to add a unit test to cover this logic.

**Review by tonyxuqqi — COMMENTED** [review]
Overall LGTM. 
update_slowness_in_store_stats is quite important and needs some test.

File: components/raftstore-v2/src/worker/pd/slowness.rs (hunk 5 above)
Comment (LykxSassinator) [reply components/raftstore-v2/src/worker/pd/slowness.rs:104]: IMO, it's redundant for it.
This logic is a compatible copy to `raftstore-v2` from `raftstore-v1`, and this test has already covered by the integration tests on `StoreHeartbeat`.

Comment (tonyxuqqi) [reply components/raftstore-v2/src/worker/pd/slowness.rs:104]: @LykxSassinator If it's covered by v2's end-to-end test, that's fine. But it seems there's no integration test for slow score in either v1 or v2. 
I think you can think to add one in a separate PR.

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**